### PR TITLE
Internal: Editor V2 - Top bar menu registration infra [ED-9952]

### PR DESCRIPTION
**Pull request by StyleShit**

### CODE REVIEW

File: packages/packages/top-bar/src/components/locations/utilities-menu-location.tsx [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
+import { getMenuId } from '../../locations';
 import { Divider } from '@elementor/ui';
 import ToolbarMenuMore from '../ui/toolbar-menu-more';
 
 const MAX_TOOLBAR_ACTIONS = 3;
 
 export default function UtilitiesMenuLocation() {
-	const injections = useInjectionsOf( LOCATION_UTILITIES_MENU_DEFAULT );
+	const injections = useInjectionsOf(
+		getMenuId( 'utilities', 'default' )
```
Comment (Nevoss) [inline packages/packages/top-bar/src/components/locations/utilities-menu-location.tsx:12]:
```suggestion
		getMenuId( 'utilities' )
```

Make the 'default' default

File: packages/packages/top-bar/src/locations/menus.tsx [hunk 2]
```diff
@@ -1,34 +1,19 @@
...
-		default: LOCATION_UTILITIES_MENU_DEFAULT,
-	},
-};
+// Will be filled using augmentation from other packages.
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface Menus {}
 
-type MenuNameMap = Readonly<typeof menuNameMap>;
-export type MenuName = keyof MenuNameMap;
+export type MenuName = keyof Menus;
```
Comment (Nevoss) [inline packages/packages/top-bar/src/locations/menus.tsx:9]:
```suggestion
export type MenuName = keyof Menus;
export type MenuGroup<MenuName> = Menus[MenuName];
```

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: packages/packages/top-bar/src/components/locations/utilities-menu-location.tsx (hunk 1 above)
Comment (StyleShit) [reply packages/packages/top-bar/src/components/locations/utilities-menu-location.tsx:12]: Done

File: packages/packages/top-bar/src/locations/menus.tsx (hunk 2 above)
Comment (StyleShit) [reply packages/packages/top-bar/src/locations/menus.tsx:9]: Done

File: packages/packages/top-bar/src/index.ts [hunk 3]
```diff
@@ -1,13 +1,22 @@
...
+export type { MenuName, MenuGroup, Menus } from './locations/menus';
+
 export * from './types';
 
+export { default as PopoverMenu } from './components/ui/popover-menu';
+export { default as PopoverMenuItem } from './components/ui/popover-menu-item';
+export { default as ToolbarMenu } from './components/ui/toolbar-menu';
+export { default as ToolbarMenuItem } from './components/ui/toolbar-menu-item';
+export { default as ToolbarMenuToggleItem } from './components/ui/toolbar-menu-toggle-item';
+
```
Comment (Nevoss) [inline packages/packages/top-bar/src/index.ts:19]: Let's be sure with @eyalnathan and @rotemee that we want to expose all of those things and make them public, and we will need to support this API.

**Review by Nevoss — DISMISSED** [review]

File: packages/packages/top-bar/src/__tests__/index.test.tsx [hunk 4]
```diff
@@ -254,59 +256,47 @@ describe( '@elementor/top-bar API', () => {
...
-		{
-			name: 'test',
-			props: {
-				title: 'Test',
-				icon: () => <span>a</span>,
-				onClick,
-			},
-		}
-	);
+function makeExampleAction( { onClick = () => null }: { onClick?: () => void } = {} ) {
```
Comment (Nevoss) [inline packages/packages/top-bar/src/__tests__/index.test.tsx:259]:
```suggestion
function mockAction( { onClick = () => null }: { onClick?: () => void } = {} ) {
```

File: packages/packages/top-bar/src/__tests__/index.test.tsx [hunk 5]
```diff
@@ -2,15 +2,14 @@ import { useState } from 'react';
 import {
```
Comment (Nevoss) [inline packages/packages/top-bar/src/__tests__/index.test.tsx:2]: Create tests for the `createMenu` API
and separate tests for the components that render the menus if needed

File: packages/packages/top-bar/src/components/locations/main-menu-location.tsx [hunk 6]
```diff
@@ -29,13 +32,13 @@ export default function MainMenuLocation() {
 				} }
 			>
 				{
-					injectionsGroups
-						.filter( ( injections ) => injections.length )
-						.map( ( injections, index ) => {
+					orderedGroups
+						.filter( ( items ) => items.length )
```
Comment (Nevoss) [inline packages/packages/top-bar/src/components/locations/main-menu-location.tsx:36]:
```suggestion
						.filter( ( group ) => group.length )
```

File: packages/packages/top-bar/src/index.ts [hunk 7]
```diff
@@ -1,13 +1,20 @@
 export {
-	registerLink,
-	registerAction,
-	registerToggleAction,
+	mainMenu,
```
Comment (Nevoss) [inline packages/packages/top-bar/src/index.ts:2]: export createMenu

File: packages/packages/top-bar/src/locations/menus.tsx [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+	const menuGroups: MenuGroup<TGroup>[] = [
+		...groups,
+		'default' as const,
+	];
+
+	const registerAction = createRegisterMenuItem( {
+		menuName,
+		menuGroups,
+		component: Action,
+	} );
```
Comment (Nevoss) [inline packages/packages/top-bar/src/locations/menus.tsx:45]:
```suggestion
	const [ ... ] = createRegisterMenuItem( {
		menuName,
		menuGroups,
		components: [ Action, ToggleAction, Link ],
	} );
```

Consider

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: packages/packages/top-bar/src/__tests__/index.test.tsx (hunk 4 above)
Comment (StyleShit) [reply packages/packages/top-bar/src/__tests__/index.test.tsx]: Done

File: packages/packages/top-bar/src/__tests__/index.test.tsx (hunk 5 above)
Comment (StyleShit) [reply packages/packages/top-bar/src/__tests__/index.test.tsx]: Done

File: packages/tests/test-utils/create-mock-menu-actions.tsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
+import { useState } from 'react';
+
+export function createMockMenuAction( { onClick = () => null }: { onClick?: () => void } = {} ) {
```
Comment (Nevoss) [inline packages/tests/test-utils/create-mock-menu-actions.tsx:3]: Add options to overwrite

File: packages/tests/test-utils/create-mock-menu-actions.tsx [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+		name: 'test',
+		props: {
+			title: 'Test',
+			icon: () => <span>a</span>,
+			onClick,
+		},
+	};
+}
+
+export function createMockMenuToggleAction() {
```
Comment (Nevoss) [inline packages/tests/test-utils/create-mock-menu-actions.tsx:14]: Add options to overwrite

File: packages/tests/test-utils/create-mock-menu-actions.tsx [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+					setClicks( ( prev ) => prev + 1 );
+				},
+				value: 'test-value',
+				disabled: clicks > 1,
+			};
+		},
+	};
+}
+
+export function createMockMenuLink() {
```
Comment (Nevoss) [inline packages/tests/test-utils/create-mock-menu-actions.tsx:36]: Add options to overwrite

File: packages/tests/test-utils/create-mock-menu-actions.tsx (hunk 9 above)
Comment (Nevoss) [reply packages/tests/test-utils/create-mock-menu-actions.tsx:3]: Add return type, so it will be clear what it retrurns

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: packages/packages/top-bar/src/components/locations/__tests__/menus.test.tsx [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+import { mainMenu, toolsMenu, utilitiesMenu } from '../../../locations';
+import { createMockMenuAction, createMockMenuLink, createMockMenuToggleAction } from 'test-utils';
+
+describe( '@elementor/top-bar - Menus components', () => {
+	describe( 'Main menu', () => {
+		it( 'should render ordered menu items in a popover', () => {
+			// Arrange.
+			mainMenu.registerAction( {
+				...createMockMenuAction(),
+				name: 'test-action',
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply packages/packages/top-bar/src/components/locations/__tests__/menus.test.tsx:14]: NO

File: packages/tests/test-utils/create-mock-menu-actions.tsx (hunk 9 above)
Comment (StyleShit) [reply packages/tests/test-utils/create-mock-menu-actions.tsx:3]: NO

File: packages/tests/test-utils/create-mock-menu-actions.tsx (hunk 10 above)
Comment (StyleShit) [reply packages/tests/test-utils/create-mock-menu-actions.tsx:14]: NO

File: packages/tests/test-utils/create-mock-menu-actions.tsx (hunk 11 above)
Comment (StyleShit) [reply packages/tests/test-utils/create-mock-menu-actions.tsx:36]: NO

**Review by Nevoss — DISMISSED** [review]

**Review by rotemee — APPROVED** [review]

**Pull request merged**